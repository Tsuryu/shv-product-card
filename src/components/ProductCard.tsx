import styles from '../styles/styles.module.css';
import { useProduct } from '../hooks/useProduct';
import React, { createContext } from 'react';
import { IProductCardProps, IProductContext } from '../interfaces/interfaces';

export const ProductContext = createContext({} as IProductContext);
const { Provider } = ProductContext;

export const ProductCard = ({
  id,
  title,
  img,
  children,
  className,
  style,
  value,
  onChange,
  initialValues,
}: IProductCardProps) => {
  const product = {
    id,
    title,
    img,
  };
  const { counter, increaseBy, maxCount, isMaxReached, reset } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        title,
        img,
        id,
        maxCount,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children instanceof Function
          ? children({
              count: counter,
              isMaxReached,
              maxCount: initialValues?.maxCount,
              product: {
                id,
                title,
                img,
              },
              increaseBy,
              reset,
            })
          : children}
      </div>
    </Provider>
  );
};
