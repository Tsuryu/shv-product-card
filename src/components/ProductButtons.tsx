import React, { useContext, CSSProperties, useCallback } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export const ProductButtons = ({
  className,
  style,
}: {
  className?: string;
  style?: CSSProperties;
}) => {
  const { counter, increaseBy, maxCount } = useContext(ProductContext);
  const isMaxReached = useCallback(() => counter === maxCount, [
    counter,
    maxCount,
  ]);

  return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}
        onClick={() => increaseBy(1)}
      >
        +
      </button>
    </div>
  );
};
