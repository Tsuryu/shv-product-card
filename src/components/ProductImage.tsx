import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';

import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

export interface IProductImage {
  className?: string;
  img?: string;
  style?: CSSProperties;
}

export const ProductImage = ({ img = '', className, style }: IProductImage) => {
  const { img: contextImage } = useContext(ProductContext);

  const imgToShow = img ? img : contextImage ? contextImage : noImage;

  return (
    <img
      src={imgToShow}
      alt="Product"
      style={style}
      className={`${styles.productImg} ${className}`}
    />
  );
};
