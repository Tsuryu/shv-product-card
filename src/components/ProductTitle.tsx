import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface IProductTitle {
  className?: string;
  title?: string;
  style?: CSSProperties;
}

export const ProductTitle = ({ title, className, style }: IProductTitle) => {
  const { title: contextTitle } = useContext(ProductContext);
  return (
    <span className={`${styles.productDescription} ${className}`} style={style}>
      {title ? title : contextTitle}
    </span>
  );
};
