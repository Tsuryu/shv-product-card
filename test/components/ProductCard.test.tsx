import * as React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src/components';
import { product1 } from '../data/products';

const { act } = renderer;

describe('ProductCard', () => {
  test('should render component', () => {
    const wrapper = renderer.create(
      <ProductCard
        title="Test product"
        className="custom class"
        id={product1.id}
      >
        {() => <h1>Product card</h1>}
      </ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('should increment counter', () => {
    const wrapper = renderer.create(
      <ProductCard
        title="Test product"
        className="custom class"
        id={product1.id}
      >
        {({ count, increaseBy }) => (
          <>
            <h1>Product card</h1>
            <span>{count}</span>
            <button onClick={() => increaseBy(2)} />
          </>
        )}
      </ProductCard>
    );

    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();

    act(() => {
      (tree as any).children[2].props.onClick();
    });
    tree = wrapper.toJSON();

    expect((tree as any).children[1].children[0]).toBe('0');
  });
});
