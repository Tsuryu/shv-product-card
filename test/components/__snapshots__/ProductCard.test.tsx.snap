// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard should increment counter 1`] = `
<div
  className="productCard custom class"
>
  <h1>
    Product card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`ProductCard should render component 1`] = `
<div
  className="productCard custom class"
>
  <h1>
    Product card
  </h1>
</div>
`;
